src/Discord.Net.SlashCommands/SlashCommandService.cs

```diff
@@ -0,0 +1,675 @@
+
+                if (_typedModuleDefs.ContainsKey(typeInfo))
+                    throw new ArgumentException("Module definition for this type already exists.");
+
+                var moduleDef = await ModuleClassBuilder.BuildAsync(new List<TypeInfo> { typeof(T).GetTypeInfo() }, this, services).ConfigureAwait(false);
+
+                if (moduleDef[typeof(T)] == default(ModuleInfo))
+                    throw new InvalidOperationException($"Could not build the module {typeInfo.FullName}, did you pass an invalid type?");
+
+                if (!_typedModuleDefs.TryAdd(typeof(T), moduleDef[typeof(T)]))
+                    throw new ArgumentException("Module definition for this type already exists.");
+
+                LoadModuleInternal(moduleDef[typeof(T)]);
+            }
+            finally
+            {
+                _lock.Release();
+            }
+        }
+
+        /// <summary>
+        /// Register/update the Application Commands to Discord
+        /// </summary>
+        /// <param name="guildId">The Id that belongs to the guild, the commands will be registered to</param>
+        /// <param name="deleteMissing">If true, delete all of the commands that are not registered in the <see cref="SlashCommandService"/></param>
+        /// <returns>A task representing the command registration process, with a collection of <see cref="RestGuildCommand"/> containing the
+        /// commands that are currently registered to the provided guild as its result.
+        /// </returns>
+        public async Task<IReadOnlyCollection<RestGuildCommand>> RegisterCommandsToGuildAsync (ulong guildId, bool deleteMissing = true)
+        {
+            EnsureClientReady();
+
+            var props = _typedModuleDefs.Values.SelectMany(x => x.ToApplicationCommandProps()).ToList();
+
+            if (!deleteMissing)
+            {
+                var existing = await ClientHelper.GetGuildApplicationCommands(Client, guildId).ConfigureAwait(false);
+                var missing = existing.Where(x => !props.Any(y => y.Name.IsSpecified && y.Name.Value == x.Name));
+                props.AddRange(missing.Select(x => x.ToApplicationCommandProps()));
+            }
+
+            return await ClientHelper.BulkOverwriteGuildApplicationCommand(Client, guildId, props.ToArray()).ConfigureAwait(false);
+        }
+
+        /// <summary>
+        /// Register/update the Application Commands to Discord
+        /// </summary>
+        /// <param name="deleteMissing">If true, delete all of the commands that are not registered in the <see cref="SlashCommandService"/></param>
+        /// <returns>A task representing the command registration process, with a collection of <see cref="RestGlobalCommand"/> containing the
+        /// global commands that are currently registered to the Discord
+        /// </returns>
+        public async Task<IReadOnlyCollection<RestGlobalCommand>> RegisterCommandsGloballyAsync (bool deleteMissing = true)
```

Comment:
To differentiate global and guild commands, maybe you can take advantage of a `[GuildCommand]` attribute or a `[GlobalCommand]` attribute.

Reply:
using attributes in this case make is so much harder to switch between production and dev environments. In most cases it's easy enough to register guild commands using the `AddModulesToGuildAsync` method. The main idea behind thi was to be able to register commands to different targets depending on the environment type, like so:
```csharp
                if (IsDebug())
                    await commands.RegisterCommandsToGuildAsync(configuration.GetValue<ulong>("testGuild"), true);
                else
                    await commands.RegisterCommandsGloballyAsync(true);
```

But maybe we can provide te users with some sort of marking mechanism to make it easier to select specific modules.